File: javascript/src/functions/security.js
// <copyright file="security.js" company="Microsoft Corporation">
// Copyright (c) Microsoft Corporation. All rights reserved.
// </copyright>

'use strict';

// Inbound Microsoft JWT validation (iss/aud/exp/RS256). Local no-auth is supported explicitly;
// on Azure it must be enabled with a pinned caller, regardless of Easy Auth headers.

const { readConfig } = require('./config');
const jwksByTenant = new Map();

function getJwks(issuerTenantId) {
    if (!jwksByTenant.has(issuerTenantId)) {
        const { createRemoteJWKSet } = require('jose');
        jwksByTenant.set(
            issuerTenantId,
            createRemoteJWKSet(new URL(`https://login.microsoftonline.com/${issuerTenantId}/discovery/v2.0/keys`)),
        );
    }
    return jwksByTenant.get(issuerTenantId);
}

// azp is the v2 caller claim, appid the v1 one.
function isExpectedCaller(payload, expectedClientId) {
    if (!expectedClientId) return true;
    const callerId = payload.azp || payload.appid;
    return typeof callerId === 'string' && callerId.toLowerCase() === expectedClientId.toLowerCase();
}

async function validateToken(request, config = readConfig()) {
    const { env, requireAuth, expectedClientId, expectedAudience: audience, tenantId, expectedIssuer } = config;
    const onAzure = !!(env.WEBSITE_INSTANCE_ID || env.WEBSITE_HOSTNAME || env.WEBSITE_SITE_NAME);
    if (onAzure && !requireAuth) {
        return { ok: false, reason: 'EPP_REQUIRE_AUTH must be true on Azure' };
    }
    if (!requireAuth) {
        return { ok: true, skipped: true };
    }

    if (onAzure && !expectedClientId) {
        return { ok: false, reason: 'EPP_EXPECTED_CLIENT_ID is required on Azure' };
    }

    if (!audience || !tenantId) {
        return { ok: false, reason: 'EPP_REQUIRE_AUTH is set but EPP_EXPECTED_AUDIENCE / EPP_TENANT_ID are missing' };
    }

    const authorizationHeader = (request.headers.get('authorization') || '').trim();
    const bearerToken = authorizationHeader.slice(0, 7).toLowerCase() === 'bearer '
        ? authorizationHeader.slice(7).trim()
        : '';
    if (!bearerToken) return { ok: false, reason: 'missing bearer token' };

    try {
        const { jwtVerify } = require('jose');
        // Accept both the v2 and v1 issuer forms unless EPP_EXPECTED_ISSUER pins one.
        const issuers = expectedIssuer
            ? [expectedIssuer]
            : [
                `https://login.microsoftonline.com/${tenantId}/v2.0`,
                `https://sts.windows.net/${tenantId}/`,
            ];
        const { payload } = await jwtVerify(bearerToken, getJwks(tenantId), {
            audience,
            issuer: issuers,
            algorithms: ['RS256'],
            requiredClaims: ['exp'],
        });

        if (!isExpectedCaller(payload, expectedClientId)) {
            return { ok: false, reason: 'unexpected caller' };
        }
        return { ok: true };
    } catch {
        return { ok: false, reason: 'token validation failed' };
    }
}

module.exports = { validateToken, isExpectedCaller };

File: python/src/security.py
import jwt
from jwt import PyJWKClient

from .config import read_config

_jwks_clients = {}


def _jwks_client(tenant_id):
    client = _jwks_clients.get(tenant_id)
    if client is None:
        client = PyJWKClient(f"https://login.microsoftonline.com/{tenant_id}/discovery/v2.0/keys")
        _jwks_clients[tenant_id] = client
    return client


def validate_token(authorization_header, config=None):
    config = read_config() if config is None else config
    require_auth = config["require_auth"]
    expected_client_id = config["expected_client_id"]
    env = config["env"]
    on_azure = bool(env.get("WEBSITE_INSTANCE_ID") or env.get("WEBSITE_HOSTNAME")
                    or env.get("WEBSITE_SITE_NAME"))
    if on_azure and (not require_auth or not (expected_client_id or "").strip()):
        return False, "auth misconfigured", None
    if not require_auth:
        return True, None, None

    audience = config["expected_audience"]
    tenant_id = config["tenant_id"]
    if not audience or not tenant_id:
        return False, "auth misconfigured", None

    if not authorization_header or not authorization_header.lower().startswith("bearer "):
        return False, "missing bearer token", None

    token = authorization_header[len("bearer "):].strip()
    if not token:
        return False, "missing bearer token", None
    try:
        signing_key = _jwks_client(tenant_id).get_signing_key_from_jwt(token)
        claims = jwt.decode(
            token,
            signing_key.key,
            algorithms=["RS256"],
            audience=audience,
            options={"verify_iss": False, "require": ["exp"]},
        )
        allowed_issuers = (
            (config["expected_issuer"],)
            if config["expected_issuer"]
            else (
                f"https://login.microsoftonline.com/{tenant_id}/v2.0",
                f"https://sts.windows.net/{tenant_id}/",
            )
        )
        if claims.get("iss") not in allowed_issuers:
            return False, "token validation failed", None

        # azp is the v2 caller claim, appid the v1 one.
        caller_id = claims.get("azp") or claims.get("appid")
        if expected_client_id:
            if not isinstance(caller_id, str) or caller_id.lower() != expected_client_id.lower():
                return False, "unexpected caller", None

        return True, None, claims.get("oid")
    except Exception:
        return False, "token validation failed", None
